File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/variable-length-quantity/variable-length-quantity.js
//
// This is only a SKELETON file for the 'Variable Length Quantity' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const encode = (numbers) => {
  const result = [];
  for (const num of numbers) {
    let n = num >>> 0;
    const chunks = [];
    do {
      const chunk = n & 0x7f;
      chunks.push(chunk);
      n = n >>> 7;
    } while (n > 0);

    chunks.reverse();

    for (let i = 0; i < chunks.length; i++) {
      let byte = chunks[i];
      if (i !== chunks.length - 1) {
        byte |= 0x80;
      }
      result.push(byte);
    }
  }
  return result;
};

export const decode = (bytes) => {
  const result = [];
  let current = 0;
  let hasData = false;

  for (const byte of bytes) {
    current = (current << 7) | (byte & 0x7f);
    hasData = true;

    if ((byte & 0x80) === 0) {
      result.push(current >>> 0);
      current = 0;
      hasData = false;
    }
  }

  if (hasData) {
    throw new Error('Incomplete sequence');
  }

  return result;
};

File: src/2025-02-24-deepclaude-rust/java/exercises/practice/variable-length-quantity/src/main/java/VariableLengthQuantity.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class VariableLengthQuantity {

    List<Long> encode(List<Long> numbers) {
        List<Long> result = new ArrayList<>();
        for (Long number : numbers) {
            long current = number;
            if (current == 0) {
                result.add(0L);
                continue;
            }
            List<Long> chunks = new ArrayList<>();
            do {
                long chunk = current & 0x7F;
                chunks.add(chunk);
                current >>>= 7;
            } while (current != 0);
            Collections.reverse(chunks);
            for (int i = 0; i < chunks.size(); i++) {
                long b = chunks.get(i);
                if (i != chunks.size() - 1) {
                    b |= 0x80;
                }
                result.add(b);
            }
        }
        return result;
    }

    List<Long> decode(List<Long> bytes) {
        List<Long> result = new ArrayList<>();
        long currentNumber = 0;
        boolean incomplete = false;
        
        for (Long b : bytes) {
            long byteValue = b & 0xFFL;
            currentNumber = (currentNumber << 7) | (byteValue & 0x7FL);
            
            if ((byteValue & 0x80L) != 0) {
                incomplete = true;
            } else {
                if (incomplete) {
                    incomplete = false;
                }
                result.add(currentNumber);
                currentNumber = 0;
            }
        }
        
        if (incomplete) {
            throw new IllegalArgumentException("Incomplete sequence");
        }
        
        return result;
    }
}
